subscription/src/main/java/org/killbill/billing/subscription/api/svcs/DefaultSubscriptionInternalApi.java

```diff
@@ -790,6 +804,67 @@ public int getDefaultBillCycleDayLocal(final Map<UUID, Integer> bcdCache, final
         }
     }
 
+    @Override
+    public UUID getAccountIdFromBundleId(final UUID bundleId, final InternalTenantContext context) throws SubscriptionBaseApiException {
+        final CacheLoaderArgument arg = createAccountIdFromBundleIdCacheLoaderArgument(context);
+        return accountIdCacheController.get(bundleId, arg);
+    }
+
+    @Override
+    public UUID getBundleIdFromSubscriptionId(final UUID subscriptionId, final InternalTenantContext context) throws SubscriptionBaseApiException {
+        final CacheLoaderArgument arg = createBundleIdFromSubscriptionIdCacheLoaderArgument(context);
+        return bundleIdCacheController.get(subscriptionId, arg);
+    }
+
+    @Override
+    public UUID getAccountIdFromSubscriptionId(final UUID subscriptionId, final InternalTenantContext context) throws SubscriptionBaseApiException {
+        final UUID bundleId = getBundleIdFromSubscriptionId(subscriptionId, context);
+        if (bundleId == null) {
+            throw new SubscriptionBaseApiException(ErrorCode.SUB_GET_NO_BUNDLE_FOR_SUBSCRIPTION, subscriptionId);
+        }
+        final UUID accountId = getAccountIdFromBundleId(bundleId, context);
+        if (accountId == null) {
+            throw new SubscriptionBaseApiException(ErrorCode.SUB_GET_INVALID_BUNDLE_ID, bundleId);
+        }
+        return accountId;
+    }
+
+    private CacheLoaderArgument createAccountIdFromBundleIdCacheLoaderArgument(final InternalTenantContext internalTenantContext) {
+        final AccountIdFromBundleIdCacheLoader.LoaderCallback loaderCallback = new AccountIdFromBundleIdCacheLoader.LoaderCallback() {
+            public UUID loadAccountId(final UUID bundleId, final InternalTenantContext internalTenantContext) {
+                final SubscriptionBaseBundle bundle;
+                try {
+                    bundle = getBundleFromId(bundleId, internalTenantContext);
+                } catch (final SubscriptionBaseApiException e) {
+                    log.warn("Unable to retrieve bundle for id='{}'", bundleId);
+                    return null;
+                }
+                return bundle.getAccountId();
+            }
+        };
+
+        final Object[] args = {loaderCallback};
+        return new CacheLoaderArgument(null, args, internalTenantContext);
+    }
+
+    private CacheLoaderArgument createBundleIdFromSubscriptionIdCacheLoaderArgument(final InternalTenantContext internalTenantContext) {
+        final BundleIdFromSubscriptionIdCacheLoader.LoaderCallback loaderCallback = new BundleIdFromSubscriptionIdCacheLoader.LoaderCallback() {
+            public UUID loadBundleId(final UUID subscriptionId, final InternalTenantContext internalTenantContext) {
+                final SubscriptionBase subscriptionBase;
+                try {
+                    subscriptionBase = getSubscriptionFromId(subscriptionId, internalTenantContext);
```

Comment:
Potentially we could create special DAO method to just return the ID. The `getSubscriptionFromId` not cheap, it needs to rehydrate all events and create the subscription state. Same remark for `getBundleFromId` and `getAccountIdFromBundleId` but less relevant as calls are cheaper.